src/main/java/com/limechain/trie/structure/TrieNodeIndex.java

```diff
@@ -0,0 +1,24 @@
+package com.limechain.trie.structure;
+
+import lombok.AccessLevel;
+import lombok.Data;
+import lombok.Setter;
+
+//TODO:
+// Consider making this data class an inner class of TrieStructure to guarantee
+// that the index has been generated by the same TrieStructure instance (with a runtime check, ofc)
```

Comment:
This idea with the "validity by construction" fell apart, since the `TrieNodeIndex` could always have been generated by one instance of the TrieStructure, but then used in another. But that's a very intentional invalid usage, so we're simply allowing the unexpected behaviour / exceptions if this happens. I think that's ok?

P.S. Also, methods in TrieStructure accepting `TrieNodeIndex` as an index are returning nullable stuff, since that's an example of "rather commonly expected value", so bothering with optionals would be too much... I think?

Reply:
If this has no other use then maybe it would be better if it's an innner class to `TrieStructure`

Reply:
> I think?

Does `TrieStructure.NodeIndex` sound good? Another reason I've moved it away is to simplify the FQN of this class (since it's a rather significant part of the public interface and outside usage), while preserving enough of the context. Also, I've sought to keep the inner body of the `TrieStructure` class minimal as to avoid ending up with a 3000 lines single class, which is also the reason to the shift of "private as package-private" (i.e. all the classes in the `trie.structure` package are semantically private parts of the `TrieStructure`, just decomposed for readability along the separation of concerns).

Reply:
You can leave it as is then.